src/com/google/enterprise/adaptor/DocIdSender.java

```diff
@@ -273,15 +332,35 @@ public GroupPrincipal pushGroupDefinitions(
     }
     journal.recordGroupPushStarted();
     String gsaVerString = config.getGsaVersion();
-    if (!new GsaVersion(gsaVerString).isAtLeast("7.2.0-0")) {
+    GsaVersion gsaVersion = new GsaVersion(gsaVerString);
+    if (!gsaVersion.isAtLeast("7.2.0-0")) {
       log.log(Level.WARNING,
-          "GSA ver {0} doesn't accept group definitions", gsaVerString);
+          "GSA ver {0} does not accept group definitions", gsaVerString);
       journal.recordGroupPushFailed();
       return defs.entrySet().iterator().next().getKey();
     }
+    if (feedType == FULL && !gsaVersion.isAtLeast("7.4.0-0")) {
+      log.log(Level.WARNING,
+          "GSA ver {0} does not support per-source replacement of all groups",
+          gsaVerString);
+      feedType = INCREMENTAL;
+    }
     if (null == handler) {
       handler = defaultErrorHandler;
     }
+    if (groupSource == null) {
+      groupSource = config.getFeedName();
+    }
+    log.log(Level.INFO, "Starting {0} push of {0} groups from source {1}",
+        new Object[] { feedType, defs.size(), groupSource });
+    String feedSourceName;
+    if (feedType == INCREMENTAL) {
+      feedSourceName = previousGroupSource(groupSource);
+    } else {
+      feedSourceName = alternateGroupSource(groupSource);
```

Comment:
One important observation is that for purposes of ACL resolution on GSA the group source is not relevant, yes?  That is, group source is relevant for managing the group source's data and potentially clearing it up wholly, but from run to run the group names and group members are the same no matter what name was used for group-source. Is that right?